src/ec/system76/ec/system76_ec.c

```diff
@@ -118,33 +123,529 @@ uint8_t system76_ec_read_version(uint8_t *data)
+		return 1;
+
+	for (i = 0; i < read_cmd[1]; i++)
+		dest[addr + i] = system76_ec_read(REG_DATA + 2 + i);
+
+	return 0;
+}
+
+/*
+ * Write len bytes from data to EC SPI bus
+ * Returns 0 on success, >0 on error
+ */
+static uint32_t ec_spi_bus_write(uint8_t *data, uint8_t len)
+{
+	uint32_t addr, i, rv;
+	uint8_t write_cmd[2] = {
+		[0] = CMD_SPI_FLAG_SCRATCH,
+		[1] = len,
+	};
+
+	for (addr = 0; addr + SPI_ACCESS_SIZE < len; addr += SPI_ACCESS_SIZE) {
+		write_cmd[1] = SPI_ACCESS_SIZE;
+
+		for (i = 0; i < write_cmd[1]; ++i)
+			system76_ec_write(REG_DATA + 2 + i, data[addr + i]);
+
+		if ((rv = system76_ec_smfi_cmd(CMD_SPI, ARRAY_SIZE(write_cmd), write_cmd)))
+			return rv;
+
+		if (system76_ec_read(REG_DATA + 1) != write_cmd[1])
+			return 1;
+	}
+
+	/* All chunks written? */
+	if (addr == len)
+		return 0;
+
+	/* Write last chunk */
+	write_cmd[1] = len % SPI_ACCESS_SIZE;
+
+	for (i = 0; i < write_cmd[1]; i++)
+		system76_ec_write(REG_DATA + 2 + i, data[addr + i]);
+
+	if ((rv = system76_ec_smfi_cmd(CMD_SPI, ARRAY_SIZE(write_cmd), write_cmd)))
+		return rv;
+
+	if (system76_ec_read(REG_DATA + 1) != write_cmd[1])
+		return 1;
+
+	return 0;
+}
+
+static int ec_spi_cmd_status(void)
+{
+	uint8_t rv;
+	uint8_t buf[1];
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	buf[0] = 0x05; /* SPI Read Status */
+
+	if ((rv = ec_spi_bus_write(buf, 1)))
+		return -rv;
+
+	if ((rv = ec_spi_bus_read(buf, 1)))
+		return -rv;
+
+	return buf[0];
+}
+
+static int ec_spi_cmd_write_enable(void)
+{
+	int status;
+	uint8_t rv;
+	uint8_t buf[1];
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	buf[0] = 0x06; /* SPI Write Enable */
+
+	if ((rv = ec_spi_bus_write(buf, 1)))
+		return -rv;
+
+	do {
+		status = ec_spi_cmd_status();
+	} while (status > 0 && (status & 3) != 2);
+
+	return 0;
+}
+
+static int ec_spi_cmd_write_disable(void)
+{
+	int status;
+	uint8_t rv;
+	uint8_t buf[1];
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	buf[0] = 0x04; /* SPI Write Disable */
+
+	if ((rv = ec_spi_bus_write(buf, 1)))
+		return -rv;
+
+	do {
+		status = ec_spi_cmd_status();
+	} while (status > 0 && (status & 3) != 0);
+
+	return 0;
+}
+
+/* Erase a sector at addr. Returns 0 on success <0 on error. */
+static long ec_spi_erase_sector(uint32_t addr)
+{
+	int status;
+	uint8_t rv;
+	uint8_t buf[4] = {
+		[0] = 0xD7, /* SPI Sector Erase */
+		[1] = (addr >> 16) & 0xFF,
+		[2] = (addr >> 8) & 0xFF,
+		[3] = addr & 0xFF,
+	};
+
+	if ((rv = ec_spi_cmd_write_enable()))
+		return -rv;
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	if ((rv = ec_spi_bus_write(buf, 4)))
+		return -rv;
+
+	do {
+		status = ec_spi_cmd_status();
+	} while (status > 0 && (status & 1) != 0);
+
+	if ((rv = ec_spi_cmd_write_disable()))
+		return -rv;
+
+	return 0;
+}
+
+/*
+ * Erase the entire chip sector by sector.
+ * Chip / bulk erase doesn't work on this hardware.
+ * Returns erased byte count on success and <0 on error.
+ */
+static long ec_spi_erase_chip(void)
```

Comment:
a/a

Reply:
Changed to int in https://github.com/Dasharo/coreboot/pull/481/commits/6165c91a2e80b65d470ce91c3fa5e0875f8fe187